### PR TITLE
State ``pkg.latest`` fails if the package is already up-to-date.

### PR DESCRIPTION
**Pull request by plastikos**
Only queue an update when:

  * package is not installed
  * package *is* installed *and* has a newer version in the repo

### What does this PR do?
Fix a bug in State ``pkg.latest`` that fails when a package is installed and up-to-date.

### What issues does this PR fix or reference?
N/A

### Previous Behavior
If a package was installed and up-to-date then ``pkg.latest`` would fail.

### New Behavior
``pkg.latest`` succeeds when a package updates correctly or it is *already* up-to-date.

### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by plastikos** [comment]
Mercurial GPG signing and Git GPG signing are completely different and ``hg-git`` doesn't translate.

That's just the way it is.

File: .hgsigs [hunk 1]
```diff
@@ -0,0 +1 @@
+d1b31775720727f9a7757865940e1ff42fba4a35 0 iQEzBAABCAAdFiEE3DYft2Ph47rtMK0vntidjI2Lq+kFAlnWt6gACgkQntidjI2Lq+kFJAgA1eP9uwqPhTgNpqluYBUr6eoaVgOi7CgYxz3sDpicdJLvrctAafvTG6eBzdpAznV4XE5O16MLuql3hO9x3oZR/Jk8jmMy4Q8cZ72849cgk4uKvvm1BNKOrJpyGf9AsTXh+/i6VReeEs2FR2QYrQ41NUDEw6EvZSOBIVE6nBvPzno09W4DTWDUKCmW2U7ny4o7fYAjtLqVY/9BT6d+CUglNlxkyCAPLnZWD8bblvmzkzJHyzkRdTBl0rAfUeuWfxhDGO4wgqujZWmBtanspT7Qntv1KnXm8xyI5t60XeZzn+ZRw/Ivf5PCHtD9iNOU7FqkbbVTkwYGebDTfizYR8ST9Q==
```
Comment (terminalmage) [inline .hgsigs]: We should not be adding this file.

Comment (plastikos) [reply .hgsigs]: Obviously not.

**Comment by plastikos** [comment]
<bump> - I deleted the ``.hgsigs`` file.

**Comment by cachedout** [comment]
@terminalmage Could you please swing back through here? Thanks.

**Review by terminalmage — COMMENTED** [review]
I can't reproduce the behavior that this is supposed to be fixing. I can't recommend a merge without more information. On which platform(s) is this happening? With which version(s) of Salt? Do you have example SLS that can be used to reproduce this?

**Comment by plastikos** [comment]
It fails on CentOS 6.x with YUM 3.2.39.

The situation happens when a package is already installed and
up-to-date and there is nothing to do.  Unfortunately the current
behavior that we observe is that it fails the ``if not
avail.get(pkg):`` test on line #2332 (because the package *is*
available)) and then falls into the ``else`` clause on line #2345 (the
else that I changed).  The ``else`` clause then adds the package as an
expected update ... which then does not happen because there is
nothing to update (it's already up-to-date).

Regardless, the "fix" simply ensures that an upgrade for the ``else``
clause only happens for two cases:

  * The package is not installed
  * The package is installed and there is a newer version in the repos

That's all this patch does - which means that it should be benign for
any existing situation.

**Comment by terminalmage** [comment]
I'm sorry, but you are mistaken. `avail` contains the return from `pkg.latest_version`. From the `pkg.latest_version` docstring:

```
    Return the latest version of the named package available for upgrade or
    installation. If more than one package name is specified, a dict of
    name/version pairs is returned.

    If the latest version of a given package is already installed, an empty
    string will be returned for that package.
```

As you can see, the return for an up-to-date package will be an empty string. Since empty strings resolve as `False`, [this `if` check will resolve as `True`](https://github.com/saltstack/salt/blob/877abb8/salt/states/pkg.py#L2332), and we will proceed to check `cur` to see if the package is currently installed (in which case we know the package is up-to-date), or does not exist.

The comments on lines 2346-2347 _are not lies_. You have simply misread the code.

Please open a bug report with a reproducible test case, and I'll be happy to investigate it. If there's an edge/corner case the existing code is not catching, I certainly want it fixed. But since this PR is based on an incorrect understanding of how the code works, I don't think it should be merged.

**Comment by rallytime** [comment]
I am going to close the PR since it hasn't received any updates or discussion for some time. Please feel free to open this again or submit a new PR when it is ready to be reviewed again. We can of course keep the conversation going below as needed. Thanks!
